typing/printtyp.ml

```diff
@@ -1638,19 +1638,37 @@ let rec tree_of_modtype ?(ellipsis=false) = function
       Omty_alias (tree_of_path Module p)
 
 and tree_of_signature sg =
-  wrap_env (fun env -> env) (tree_of_signature_rec !printing_env false) sg
-
-and tree_of_signature_rec env' in_type_group = function
-    [] -> stop_type_group env'; []
-  | item :: rem as items ->
-      let in_type_group = still_in_type_group env' in_type_group item in
-      let (sg, rem) = filter_rem_sig item rem in
-      hide_rec_items items;
-      protect_rec_items items;
-      reset_naming_context ();
-      let trees = trees_of_sigitem item in
-      let env' = Env.add_signature (item :: sg) env' in
-      trees @ tree_of_signature_rec env' in_type_group rem
+  wrap_env (fun env -> env)
+    (tree_of_signature_rec (fun _ (_,x) -> x) !printing_env) sg
+
+and tree_of_signature_rec : 'r. (_->_->'r) -> _ -> _ -> 'r list
+  = fun decorate env' sg ->
+  let structured = group_recursive_items (group_syntactic_items sg) in
+  let collect_trees_of_rec_group (env,trees) group =
+    let env', more_trees = trees_of_recursive_sigitem_group env group in
+    let more_trees = List.map (decorate env) more_trees in
+    set_printing_env env';
+    env', more_trees :: trees in
+  let _, rev_trees =
+    List.fold_left collect_trees_of_rec_group (env',[]) structured in
+  List.flatten (List.rev rev_trees)
+
+and trees_of_recursive_sigitem_group env group =
+  let display x =
+    reset_naming_context ();
+    List.map (fun y -> x.src,y) (trees_of_sigitem x.src) in
+  match group with
+  | Not_rec x -> add_sigitem env x, display x
+  | Rec_group (ids,items) ->
+      let is_type = match items with
+        | {src=Sig_type _; _ } :: _ -> true
+        | _ -> false  in
```

Comment:
This code makes the assumption (compared to the previous code) that if one item of a recursive group is a `Sig_type` declaration, then all items are types. Are we sure this is true? (I *think* it currently is true.) One could also *not* make this assumption by collecting the `ids` to hide with a `filter_map`.

Reply:
Class recursive group can contains types. At the end, it was simpler to be less clever here and let the hidding functions handle this part.